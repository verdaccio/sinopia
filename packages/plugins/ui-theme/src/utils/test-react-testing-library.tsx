import { render } from '@testing-library/react';
import React from 'react';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';

import ThemeProvider from 'verdaccio-ui/design-tokens/ThemeProvider';
import APIProvider from 'verdaccio-ui/providers/API/APIProvider';
import AppConfigurationProvider from 'verdaccio-ui/providers/config';

import i18nConfig from '../i18n/config';

const renderWithStore = (ui, store) =>
  render(ui, {
    wrapper: ({ children }) => (
      <Provider store={store}>
        <AppConfigurationProvider>
          <APIProvider>
            <ThemeProvider>
              <I18nextProvider i18n={i18nConfig}>{children}</I18nextProvider>
            </ThemeProvider>
          </APIProvider>
        </AppConfigurationProvider>
      </Provider>
    ),
  });

const customRender = (node: React.ReactElement, ...options: any) => {
  return render(
    <AppConfigurationProvider>
      <APIProvider>
        <ThemeProvider>
          <I18nextProvider i18n={i18nConfig}>{node}</I18nextProvider>
        </ThemeProvider>
      </APIProvider>
    </AppConfigurationProvider>,
    ...options
  );
};

export * from '@testing-library/react';
// FIXME: rename all references with customRemder
export { customRender as render };
export { customRender, renderWithStore };
