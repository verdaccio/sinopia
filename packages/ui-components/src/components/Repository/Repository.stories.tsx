import type { Meta, StoryObj } from '@storybook/react';
import React from 'react';

import { default as Repository } from '.';

const meta: Meta<typeof Repository> = {
  title: 'Components/Sidebar/Repository',
  component: Repository,
};
export default meta;
type Story = StoryObj<typeof Repository>;

export const Primary: Story = {
  name: 'Git Repository',
  render: () => {
    return (
      <Repository
        packageMeta={{
          _uplinks: {},
          latest: {
            name: 'verdaccio-ui/local-storage',
            version: '8.0.1-next.1',
            repository: {
              type: 'git',
              url: 'git://github.com/verdaccio/ui.git',
            },
          },
        }}
      />
    );
  },
};

export const HTTPS: Story = {
  name: 'Https repo',
  render: () => (
    <Repository
      packageMeta={{
        _uplinks: {},
        latest: {
          name: 'verdaccio-ui/local-storage',
          version: '8.0.1-next.1',
          repository: {
            type: 'https',
            url: 'https://github.com/verdaccio/ui.git',
          },
        },
      }}
    />
  ),
};

export const HTTP: Story = {
  name: 'Http repo',
  render: () => (
    <Repository
      packageMeta={{
        _uplinks: {},
        latest: {
          name: 'verdaccio-ui/local-storage',
          version: '8.0.1-next.1',
          repository: {
            type: 'http',
            url: 'http://github.com/verdaccio/ui.git',
          },
        },
      }}
    />
  ),
};
