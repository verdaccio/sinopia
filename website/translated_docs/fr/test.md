---
id: unit-testing
title: "Test unitaire"
---
Tous les tests sont divisés en trois dossiers:

- `test/unit` - Tests couvrant les fonctions transformant les données de manière non triviale. Ces tests `demandent()` simplement peu de fichiers et exécutent le code qu'ils contiennent, ils sont donc très rapides.
- `test/fonctionnel` - Tests qui lancent une instance verdaccio et effectuent une série de requêtes via http. Ils sont plus lents que les tests unitaires.
- `test/integration` - Tests qui lancent une instance de verdaccio et lui adressent des requêtes à l'aide de npm. Ils sont considérablement lents et peuvent atteindre un registre npm réel. **test non maintenu**

Les tests unitaires et fonctionnels sont effectués automatiquement lorsque vous démarrez `npm test` à partir du dossier racine du projet. Les tests d'intégration doivent être effectués manuellement de temps à autre.

L'on utilise `jest` pour tous les tests.

## Le Script npm

Pour exécuter le script de test, vous pouvez utiliser soit `npm` ou `yarn`.

    yarn run test
    

Cela activera uniquement les deux premiers groupes de test, unité et fonctionnel.

### Utilisation de test/unit

Ce qui suit est juste un exemple de ce à quoi un test unitaire devrait ressembler. Suivez fondamentalement la norme `jest`.

Essayez de décrire ce que fait exactement le test unitaire en une seule phrase dans l'en-tête de la section `test`.

```javacript
const verdaccio = require('../../src/api/index');
const config = require('./partials/config');

describe('basic system test', () => {

  beforeAll(function(done) {
    // something important
  });

  afterAll((done) => {
    // undo something important
  });

  test('server should respond on /', done => {
    // your test
    done();
  });
});
```

### Utilisation de test/functional

Le test fonctionnel de verdaccio présente un niveau de complexité supérieur, qui nécessite une explication détaillée pour garantir une expérience positive.

Tout commence dans le fichier de`index.js`. Plongeons-nous dedans.

```javascript
// we create 3 server instances
 const config1 = new VerdaccioConfig(
    './store/test-storage',
    './store/config-1.yaml',
    'http://localhost:55551/');
  const config2 = new VerdaccioConfig(
      './store/test-storage2',
      './store/config-2.yaml',
      'http://localhost:55552/');
  const config3 = new VerdaccioConfig(
        './store/test-storage3',
        './store/config-3.yaml',
        'http://localhost:55553/');
  const server1: IServerBridge = new Server(config1.domainPath);
  const server2: IServerBridge = new Server(config2.domainPath);
  const server3: IServerBridge = new Server(config3.domainPath);
  const process1: IServerProcess = new VerdaccioProcess(config1, server1, SILENCE_LOG);
  const process2: IServerProcess = new VerdaccioProcess(config2, server2, SILENCE_LOG);
  const process3: IServerProcess = new VerdaccioProcess(config3, server3, SILENCE_LOG);
  const express: any = new ExpressServer();
  ...

    // we check whether all instances has been started, since run in independent processes
    beforeAll((done) => {
      Promise.all([
        process1.init(),
        process2.init(),
        process3.init()]).then((forks) => {
          _.map(forks, (fork) => {
            processRunning.push(fork[0]);
          });
          express.start(EXPRESS_PORT).then((app) =>{
            done();
          }, (err) => {
            done(err);
          });
      }).catch((error) => {
        done(error);
      });
    });

    // after finish all, we ensure are been stoped
    afterAll(() => {
      _.map(processRunning, (fork) => {
        fork.stop();
      });
      express.server.close();
    });


```

### Utilisation

Ici nous allons décrire à quoi devrait ressembler un test fonctionnel typique. Vérifiez inline pour plus d'informations détaillées.

#### The lib/server.js

La classe de serveur est uniquement un wrapper qui simule un client `npm` et fournit une simple API pour les tests fonctionnels.

Comme nous l'avons mentionné dans la section précédente, nous créons 3 serveurs de processus accessibles dans chaque processus, tels que `serveur1`, `serveur2` et `` serveur3`.

En utilisant ces références, vous pourrez envoyer des requêtes à chacune des 3 instances en cours d'exécution.

```javascript
<br />export default function(server) {
  // we recieve any server instance via arguments
  test('add tag - 404', () => {
    // we interact with the server instance.
    return server.addTag('testpkg-tag', 'tagtagtag', '0.0.1').status(404).body_error(/no such package/);
  });
});
```

### Test/integration

Cette section n'a jamais été utilisée, mais nous recherchons de l'aide pour la faire fonctionner correctement. **Toute nouvelle idée est la bienvenue.**