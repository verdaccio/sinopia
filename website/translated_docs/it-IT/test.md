---
id: unit-testing
title: "Test Unità"
---
Verdaccio si concentra molto sui test delle unità e sui test end to end. Se sei disposto a contribuire, fai riferimento alla seguente [guida per maggiori informazioni](https://github.com/verdaccio/verdaccio/wiki/Developing-new-tests).