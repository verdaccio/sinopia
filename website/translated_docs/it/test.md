---
id: unit-testing
title: "Test Unità"
---
Tutti i test sono divisi in tre cartelle:

- `test/unit` - Test che coprono le funzioni che trasformano i dati in modo non banale. Questi test `require()` necessitano semplicemente alcuni file ed eseguono il codice in essi, quindi sono molto rapidi.
- `test/functional` - Test che lanciano un'istanza di verdaccio e realizzano una serie di richieste a essa su http. Sono più lenti dei test unità.
- `test/integration` - Test che lanciano un'istanza di verdaccio e fanno richieste a essa usando npm. Sono considerevolmente lenti e possono raggiungere un registro npm reale. **test non mantenuto**

I test unità e funzionali vengono eseguiti automaticamente all'avvio di `npm test` dalla cartella principale del progetto. I test di integrazione vanno eseguiti manualmente di volta in volta.

Usiamo `jest` per tutti i test.

## Lo Script npm

Per eseguire lo script del test si può utilizzare sia `npm` che `yarn`.

    yarn run test
    

Questo azionerà solo i primi due gruppi di test, unità e funzionale.

### Utilizzo di test/unit

Il seguente è solo un esempio di come dovrebbe vedersi un test unità. Fondamentalmente seguire il `jest` standard.

Provare a descrivere cosa fa esattamente il test unità in una sola frase nell'intestazione della sezione `test`.

```javacript
const verdaccio = require('../../src/api/index');
const config = require('./partials/config');

describe('basic system test', () => {

  beforeAll(function(done) {
    // something important
  });

  afterAll((done) => {
    // undo something important
  });

  test('server should respond on /', done => {
    // your test
    done();
  });
});
```

### Utilizzo di test/functional

Il test funzionale in verdaccio presenta un maggior livello di complessità che necessita di una spiegazione esaustiva per assicurarti un'esperienza positiva.

Tutto inizia nel file `index.js`. Andiamo ad analizzarlo da vicino.

```javascript
// we create 3 server instances
 const config1 = new VerdaccioConfig(
    './store/test-storage',
    './store/config-1.yaml',
    'http://localhost:55551/');
  const config2 = new VerdaccioConfig(
      './store/test-storage2',
      './store/config-2.yaml',
      'http://localhost:55552/');
  const config3 = new VerdaccioConfig(
        './store/test-storage3',
        './store/config-3.yaml',
        'http://localhost:55553/');
  const server1: IServerBridge = new Server(config1.domainPath);
  const server2: IServerBridge = new Server(config2.domainPath);
  const server3: IServerBridge = new Server(config3.domainPath);
  const process1: IServerProcess = new VerdaccioProcess(config1, server1, SILENCE_LOG);
  const process2: IServerProcess = new VerdaccioProcess(config2, server2, SILENCE_LOG);
  const process3: IServerProcess = new VerdaccioProcess(config3, server3, SILENCE_LOG);
  const express: any = new ExpressServer();
  ...

    // we check whether all instances has been started, since run in independent processes
    beforeAll((done) => {
      Promise.all([
        process1.init(),
        process2.init(),
        process3.init()]).then((forks) => {
          _.map(forks, (fork) => {
            processRunning.push(fork[0]);
          });
          express.start(EXPRESS_PORT).then((app) =>{
            done();
          }, (err) => {
            done(err);
          });
      }).catch((error) => {
        done(error);
      });
    });

    // after finish all, we ensure are been stoped
    afterAll(() => {
      _.map(processRunning, (fork) => {
        fork.stop();
      });
      express.server.close();
    });


```

### Utilizzo

Qui andremo a descrivere come dovrebbe apparire un test funzionale tipico.

#### Il lib/server.js

La classe server è solamente un wrapper che simula un `npm` client e fornisce un API semplice per il test funzionale.

Come abbiamo menzionato nella sezione precedente, stiamo creando 3 server di processo che siano accessibili in ogni processo come `server1`, `server2` e ``server3`.

Usando queste referenze sarai in grado di inviare richieste a ognuna delle 3 istanze in esecuzione.

```javascript
<br />export default function(server) {
  // we recieve any server instance via arguments
  test('add tag - 404', () => {
    // we interact with the server instance.
    return server.addTag('testpkg-tag', 'tagtagtag', '0.0.1').status(404).body_error(/no such package/);
  });
});
```

### Test/integration

Questa sezione non è ancora mai stata usata, ma stiamo cercando aiuto per farla funzionare a dovere. **Ogni nuova idea è benvenuta**