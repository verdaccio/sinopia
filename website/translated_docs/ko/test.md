---
id: unit-testing
title: "유닛 테스트"
---
Verdaccio has strong focus on unit testing and end to end testing. If you are willing to contribute, please refer to the following [guide for more information](https://github.com/verdaccio/verdaccio/wiki/Developing-new-tests).