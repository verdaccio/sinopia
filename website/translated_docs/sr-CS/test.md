---
id: unit-testing
title: "Unit Testing"
---
Svi testovi su podeljeni u tri foldera:

- `test/unit` - Testovi koji pokrivaju funkcije koje transformišu podatke na netrivijalni način. Da biste testirali, koristite `require()` na nekolliko fajlova i pokrenite kod, brzo se izvršava.
- `test/functional` - Test koji pokreće verdaccio instancu i pokreće seriju zahteva nad http. Ovaj tip testova je sporiji od unit testova.
- `test/integration` - Test koji pokreće verdaccio instancu i zahteva je koristeći npm. Ovaj tip testova je izuzetno spor i može da pogodi pravi npm registry. **unmaintained test**

Unit i functional testovi se automatski izvršavaju pokretanjem `npm test` iz root directorijuma projekta. Integration testovi bi trebalo da se pokreću ručno, s vremena na vreme.

Koristimo `jest` za sve testove.

## The npm Script

Da biste pokrenuli test skriptu, možete koristiti bilo `npm` ili `yarn`.

    yarn run test
    

To će pokrenuti samo prve dve grupe testova, unit i functional.

### Korišćenje test/unit

U navedenom primeru možete videti kako bi unit test trebalo da izgleda. Suštinski, samo pratite `jest` standard.

Pokušajte da u jednoj rečenici u zaglavlju `test` sekcije objasnite tačno šta bi unit test trebalo da radi.

```javacript
const verdaccio = require('../../src/api/index');
const config = require('./partials/config');

describe('basic system test', () => {

  beforeAll(function(done) {
    // something important
  });

  afterAll((done) => {
    // undo something important
  });

  test('server should respond on /', done => {
    // your test
    done();
  });
});
```

### Korišćenje test/functional

Funkcionalno testiranje u verdaccio-u je nešto kompleksnije i zahteva detaljno objašnjenje.

Sve počinje od `index.js` fajla. Hajde da se bacimo u rešavanje problema.

```javascript
// we create 3 server instances
 const config1 = new VerdaccioConfig(
    './store/test-storage',
    './store/config-1.yaml',
    'http://localhost:55551/');
  const config2 = new VerdaccioConfig(
      './store/test-storage2',
      './store/config-2.yaml',
      'http://localhost:55552/');
  const config3 = new VerdaccioConfig(
        './store/test-storage3',
        './store/config-3.yaml',
        'http://localhost:55553/');
  const server1: IServerBridge = new Server(config1.domainPath);
  const server2: IServerBridge = new Server(config2.domainPath);
  const server3: IServerBridge = new Server(config3.domainPath);
  const process1: IServerProcess = new VerdaccioProcess(config1, server1, SILENCE_LOG);
  const process2: IServerProcess = new VerdaccioProcess(config2, server2, SILENCE_LOG);
  const process3: IServerProcess = new VerdaccioProcess(config3, server3, SILENCE_LOG);
  const express: any = new ExpressServer();
  ...

    // we check whether all instances has been started, since run in independent processes
    beforeAll((done) => {
      Promise.all([
        process1.init(),
        process2.init(),
        process3.init()]).then((forks) => {
          _.map(forks, (fork) => {
            processRunning.push(fork[0]);
          });
          express.start(EXPRESS_PORT).then((app) =>{
            done();
          }, (err) => {
            done(err);
          });
      }).catch((error) => {
        done(error);
      });
    });

    // after finish all, we ensure are been stoped
    afterAll(() => {
      _.map(processRunning, (fork) => {
        fork.stop();
      });
      express.server.close();
    });


```

### Korišćenje

Ovde ćemo pokazati kako izgleda funkcionalni test, proverite inline i saznajte više detalja.

#### The lib/server.js

Server class je samo wrapper koji simulira `npm` client i obezbeđuje jednostavan API za funkcionalni test.

Kao što smo već opisali u prethodnoj sekciji, kreiramo 3 proces servera koji su dostupni u svakom procesu kao `server1`, `server2` i``server3`.

Koristeći takve reference, moći ćete da šaljete zahtev do bilo koje 3 pokrenute instance.

```javascript
<br />export default function(server) {
  // we recieve any server instance via arguments
  test('add tag - 404', () => {
    // we interact with the server instance.
    return server.addTag('testpkg-tag', 'tagtagtag', '0.0.1').status(404).body_error(/no such package/);
  });
});
```

### Test/integration

Ove sekcije nikada nisu korišćene, te stoga tražimo pomoć kako bi radile na ispravan način. **Svaka ideja i svaka pomoć je dobrodošla.**