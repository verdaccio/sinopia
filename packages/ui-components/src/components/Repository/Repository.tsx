import styled from '@emotion/styled';
import Avatar from '@mui/material/Avatar';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import ListItemText from '@mui/material/ListItemText';
import Typography from '@mui/material/Typography';
import React from 'react';
import { useTranslation } from 'react-i18next';

import { Theme } from '../../Theme';
import { url as urlUtils } from '../../utils';
import CopyClipboard from '../CopyClipboard';
import { Git } from '../Icons';
import { Link } from '../Link';

const StyledText = styled(Typography)<{ theme?: Theme }>((props) => ({
  fontWeight: props.theme?.fontWeight.bold,
  textTransform: 'capitalize',
}));

const GithubLink = styled(Link)<{ theme?: Theme }>(({ theme }) => ({
  color: theme?.palette.mode === 'light' ? theme?.palette.primary.main : theme?.palette.white,
  ':hover': {
    color: theme?.palette.dodgerBlue,
  },
}));

const RepositoryListItem = styled(ListItem)({
  padding: 0,
  ':hover': {
    backgroundColor: 'transparent',
  },
});

const RepositoryListItemText = styled(ListItemText)({
  padding: '0 10px',
  margin: 0,
});

const RepositoryAvatar = styled(Avatar)({
  borderRadius: '0px',
  padding: '0',
  backgroundColor: 'transparent',
});

const Repository: React.FC<{ packageMeta: any }> = ({ packageMeta }) => {
  const { t } = useTranslation();
  const url = packageMeta?.latest?.repository?.url;
  if (!url || !urlUtils.isURL(url)) {
    return null;
  }

  const getCorrectRepositoryURL = (): string => {
    if (!url.includes('git+')) {
      return url;
    }

    return url.split('git+')[1];
  };

  const repositoryURL = getCorrectRepositoryURL();

  return (
    <List
      dense={true}
      subheader={<StyledText variant="subtitle1">{t('sidebar.repository.title')}</StyledText>}
    >
      <RepositoryListItem>
        <RepositoryAvatar sx={{ backgroundColor: '#fff' }}>
          <Git />
        </RepositoryAvatar>
        <RepositoryListItemText
          primary={
            <CopyClipboard dataTestId="repositoryID" text={repositoryURL} title={repositoryURL}>
              <GithubLink external={true} to={repositoryURL} variant="outline">
                {repositoryURL}
              </GithubLink>
            </CopyClipboard>
          }
        />
      </RepositoryListItem>
    </List>
  );
};

export default Repository;
