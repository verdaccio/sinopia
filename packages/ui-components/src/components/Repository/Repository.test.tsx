import React from 'react';

import { render } from '../../test/test-react-testing-library';
import Repository from './Repository';
import data from './__partials__/data.json';

describe('<Repository /> component', () => {
  test('should load the component in default state', () => {
    const { container } = render(<Repository packageMeta={data} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render the component in with no repository data', () => {
    const packageMeta = {
      ...data,
      latest: {
        ...data?.latest,
        repository: undefined,
      },
    };

    const { queryByText } = render(<Repository packageMeta={packageMeta} />);

    expect(queryByText('Repository')).toBeFalsy();
  });

  test('should render the component in with invalid url', () => {
    const packageMeta = {
      ...data,
      latest: {
        ...data?.latest,
        repository: {
          type: 'git',
          url: 'git://github.com/verdaccio/ui.git',
        },
      },
    };

    const { queryByText } = render(<Repository packageMeta={packageMeta} />);

    expect(queryByText('Repository')).toBeFalsy();
  });
});
