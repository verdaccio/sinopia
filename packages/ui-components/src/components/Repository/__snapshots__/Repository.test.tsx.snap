// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Repository /> component > should load the component in default state 1`] = `null`;
